### PR TITLE
Complete WebAssembly/WASI Support

### PR DESCRIPTION
**Pull request by kateinoigakukun**
### Motivation:

Trying to support WebAssembly/WASI before shipping swift-testing as a part of toolchain

### Modifications:

Thanks to prior WASI compatibility efforts by @grynspan, most of the parts work as is now. Here are a few items I changed:

- Remove `SWT_NO_FILE_IO` definition since wasi-libc provides all IO functionalities we need
- Stop including `signal.h`
- Fix metadata accessor function pointer representation on WASI to use absolute pointer instead of relative pointer
- Define primitive lock type as `Void` for WASI
  It's not the optimal representation but I couldn't find a better way 😓 

### Result:

Tested with `DEVELOPMENT-SNAPSHOT-2024-07-15-a` toolchain

```console
$ swift sdk install https://github.com/swiftwasm/swift/releases/download/swift-wasm-DEVELOPMENT-SNAPSHOT-2024-07-16-a/swift-wasm-DEVELOPMENT-SNAPSHOT-2024-07-16-a-wasm32-unknown-wasi.artifactbundle.zip

$ swift package init --name Hello --enable-experimental-swift-testing
# Replace swit-testing dependency entry in Package.swift with:
#   .package(url: "https://github.com/kateinoigakukun/swift-testing.git", branch: "yt/wasi-support"),

$ swift build --swift-sdk DEVELOPMENT-SNAPSHOT-2024-07-16-a-wasm32-unknown-wasi --build-tests --enable-experimental-swift-testing
$ wasmkit-cli run --dir . ./.build/debug/HelloPackageTests.wasm
◇ Test run started.
↳ Testing Library Version: 3702b3a53779ccca0f8205b7a76a6f5038197120
◇ Test example() started.
✔ Test example() passed after 0.041244319 seconds.
✔ Test run with 1 test passed after 0.081816901 seconds.
```

### Checklist:

- [x] Code and documentation should follow the style of the [Style Guide](https://github.com/apple/swift-testing/blob/main/Documentation/StyleGuide.md).
- [x] If public symbols are renamed or modified, DocC references should be updated.

### CODE REVIEW

**Comment by kateinoigakukun** [comment]
@swift-ci test

**Review by MaxDesiatov — APPROVED** [review]

File: Sources/_TestingInternals/include/Includes.h
Comment (grynspan) [inline Sources/_TestingInternals/include/Includes.h:30]: What goes wrong if signal.h is included on WASI? Is it harmful?

**Comment by grynspan** [comment]
Thanks for taking a look at all this! We really appreciate it. :)

File: Sources/_TestingInternals/include/Includes.h
Comment (grynspan) [reply Sources/_TestingInternals/include/Includes.h:30]: I see you commented about this in the commit description. Please document the issue in a comment here so we can refer back to it in the future. Thanks!

File: Package.swift
Comment (grynspan) [reply Package.swift:133]: At an educated guess, most `os(Linux)` branches in FileHandle.swift should also be used by WASI. :)

Comment (kateinoigakukun) [reply Package.swift:133]: Oh, good catch. I missed them!

Comment (grynspan) [reply Package.swift:133]: I'd search the package for any uses of `SWT_NO_FILE_IO` and sanity-check 'em too.

File: Sources/Testing/Support/Locked.swift
Comment (grynspan) [reply Sources/Testing/Support/Locked.swift:45]: Note I have adoption of `Mutex` on a branch here: https://github.com/apple/swift-testing/tree/jgrynspan/adopt-mutex

It defines the "no supported lock type" case to `Never`, but I'd prefer to defer to `Mutex` and let it decide to be a no-op if possible.

I'm not planning to merge that branch in the near future, but it's relevant.

File: Sources/_TestingInternals/Discovery.cpp
Comment (grynspan) [inline Sources/_TestingInternals/Discovery.cpp:96]: This `static_assert()` can just be on one line, I think. Code doesn't need to wrap at 80 columns, just comments.

File: Sources/_TestingInternals/Discovery.cpp
Comment (grynspan) [inline Sources/_TestingInternals/Discovery.cpp:93]: Is the mask value used? If not, does this type need to be generic over it? (I assume yes, for the sake of calling code?)

File: Sources/_TestingInternals/Discovery.cpp
Comment (kateinoigakukun) [reply Sources/_TestingInternals/Discovery.cpp:123]: Yes, we still use `SWTRelativePointer` to reference a data space on wasm32. We need absolute pointer just for references to code space.

Comment (grynspan) [reply Sources/_TestingInternals/Discovery.cpp:123]: Okay! Makes sense.

File: Sources/_TestingInternals/Discovery.cpp
Comment (kateinoigakukun) [reply Sources/_TestingInternals/Discovery.cpp:93]: At least for now, there is no metadata that embeds payload within a function pointer, so we actually don't use the mask value

**Comment by kateinoigakukun** [comment]
@swift-ci test

File: Sources/_TestingInternals/include/Includes.h
Comment (kateinoigakukun) [reply Sources/_TestingInternals/include/Includes.h:34]: Added a comment there :)

**Comment by grynspan** [comment]
One other thing! Can you modify the project readme to include this line in the supported OS table? 😁

```md
| **Wasm** | | | Experimental |
```

File: Sources/Testing/Support/Locked.swift
Comment (grynspan) [reply Sources/Testing/Support/Locked.swift:45]: Perfect!

File: Sources/Testing/Support/FileHandle.swift
Comment (grynspan) [inline Sources/Testing/Support/FileHandle.swift:376]: Can we assume WASI doesn't support TTYs, or is it possible for this to come back `true`?

File: Sources/Testing/Traits/Tags/Tag.Color+Loading.swift
Comment (grynspan) [inline Sources/Testing/Traits/Tags/Tag.Color+Loading.swift:52]:
```suggestion
/// The value of this property is `nil` if the platform does not support the concept of a home directory, or if the home directory could not be determined.
```

(Wrap to 80 columns.)

File: Sources/Testing/Traits/Tags/Tag.Color+Loading.swift
Comment (grynspan) [inline Sources/Testing/Traits/Tags/Tag.Color+Loading.swift:62]:
```suggestion
  return nil
```

File: Sources/Testing/Traits/Tags/Tag.Color+Loading.swift
Comment (grynspan) [inline Sources/Testing/Traits/Tags/Tag.Color+Loading.swift:67]:
```suggestion
  return nil
```

File: Sources/Testing/Support/FileHandle.swift
Comment (kateinoigakukun) [reply Sources/Testing/Support/FileHandle.swift:376]: wasi-libc provides `isatty` implementation (not a stub!), so it can be `true` https://github.com/WebAssembly/wasi-libc/blob/main/libc-bottom-half/sources/isatty.c

File: Sources/Testing/Traits/Tags/Tag.Color+Loading.swift
Comment (grynspan) [inline Sources/Testing/Traits/Tags/Tag.Color+Loading.swift:94]:
```suggestion
    // If the platform does not support user-specific configuration, skip custom
    // tag colors.
```

File: Sources/Testing/Support/FileHandle.swift
Comment (MaxDesiatov) [reply Sources/Testing/Support/FileHandle.swift:376]: You can read from a WASI-compatible module's stdout, so it's a matter of the consumer of that stdout whether they support terminal emulation or not. There are definitely in-browser terminal emulation libraries, and a Wasm runtime itself could be running as CLI in a terminal emulator. IMO seems premature to exclude WASI from that.

File: Sources/Testing/Support/FileHandle.swift
Comment (grynspan) [inline Sources/Testing/Support/FileHandle.swift:402]: As with TTYs, is it actually possible for a file to be a pipe?

Comment (kateinoigakukun) [reply Sources/Testing/Support/FileHandle.swift:402]: WASI Preview 1 doesn't have fifo nor pipe but Preview 2 does! https://github.com/WebAssembly/wasi-filesystem/blob/fe0ebbcb8bab1ee1f860812c25a6cc83b725e47b/wit/types.wit#L52

**Comment by grynspan** [comment]
@swift-ci please test

File: Sources/Testing/Support/FileHandle.swift
Comment (grynspan) [reply Sources/Testing/Support/FileHandle.swift:402]: Okay, works for me!

File: Sources/Testing/Support/FileHandle.swift
Comment (grynspan) [reply Sources/Testing/Support/FileHandle.swift:376]: @MaxDesiatov I want to avoid accidentally opting us into a state where the output is garbled. If we're not certain a feature works properly, it's better to disable it now and come back to it later than to preemptively enable it.

Pipes are problematic in this way because you can run a series of pipes that ultimately ends at a terminal emulator that supports ANSI escape codes, but actually determining whether outputtting ANSI escape codes is one of the Hard Problems™ in Computer Science™.

But, since @kateinoigakukun pointed out that `isatty()` is fully implemented, we can just defer to it. (Adding a flag to enable writing escape codes to an arbitrary file would be out of the scope of this PR, but we can chat about it in the forums.)

**Review by grynspan — APPROVED** [review]
:shipit:

**Comment by grynspan** [comment]
Ready to merge when you are.

**Comment by kateinoigakukun** [comment]
⛵

**Pull request merged**
